include/dice/template-library/channel.hpp

```diff
@@ -0,0 +1,304 @@
+#ifndef DICE_TEMPLATELIBRARY_CHANNEL_HPP
+#define DICE_TEMPLATELIBRARY_CHANNEL_HPP
+
+#include <atomic>
+#include <cassert>
+#include <condition_variable>
+#include <cstddef>
+#include <deque>
+#include <mutex>
+#include <optional>
+
+namespace dice::template_library {
+
+	/**
+     * A single producer, single consumer channel/queue
+     * @note this can technically be used as a multi producer, single consumer queue, but care must be taken
+     *      of when exactly close() is called
+     * @warning close() must be called once the producing thread is done, otherwise the reading thread will hang indefinitely or SEGFAULT
+     *
+     * @tparam T value type of the channel
+     */
+    template<typename T>
+    struct channel {
+        using value_type = T;
+        using size_type = size_t;
+        using difference_type = std::ptrdiff_t;
+        using reference = value_type &;
+        using const_reference = value_type const &;
+        using pointer = T *;
+        using const_pointer = T const *;
+
+    private:
+        size_t max_cap_; ///< maximum allowed number of elements in queue_
+        std::deque<T> queue_; ///< queue for elements
```

Comment:
```suggestion
        std::deque<T> queue_ = ATOMIC_FLAG_INIT; ///< queue for elements
```

Was not default initialized before C++20. So this would be safer. Alternatively, preprocessor guards could be used to have a pre- and from-C++20 version